core/src/main/java/org/sonarsource/sonarlint/core/log/MessageFormat.java

```diff
@@ -0,0 +1,490 @@
+  // 11 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeEscapedEscapeChars(final int escapeCounter, final StringBuilder buffer) {
+    final int escapedEscapes = escapeCounter >> 1; // divide by two
+    writeUnescapedEscapeChars(escapedEscapes, buffer);
+  }
+
+  /**
+   * Writes the specified number of '\' chars to the specified result array (starting at the specified position) and
+   * returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 20 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeUnescapedEscapeChars(int escapeCounter, final StringBuilder buffer) {
+    while (escapeCounter > 0) {
+      buffer.append(ESCAPE_CHAR);
+      escapeCounter--;
+    }
+  }
+
+  /**
+   * Appends the argument at the specified argument index (or, if no such argument exists, the "{}" delimiter pair) to
+   * the specified result char array at the specified position and returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 25 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeArgOrDelimPair(final Object[] arguments, final int argCount, final int currentArgument,
+    final StringBuilder buffer) {
+    if (currentArgument < argCount) {
+      recursiveDeepToString(arguments[currentArgument], buffer, null);
+    } else {
+      writeDelimPair(buffer);
+    }
+  }
+
+  /**
+   * This method performs a deep toString of the given Object.
+   * Primitive arrays are converted using their respective Arrays.toString methods while
+   * special handling is implemented for "container types", i.e. Object[], Map and Collection because those could
+   * contain themselves.
+   * <p>
+   * dejaVu is used in case of those container types to prevent an endless recursion.
+   * </p>
+   * <p>
+   * It should be noted that neither AbstractMap.toString() nor AbstractCollection.toString() implement such a
+   * behavior.
+   * They only check if the container is directly contained in itself, but not if a contained container contains the
+   * original one. Because of that, Arrays.toString(Object[]) isn't safe either.
+   * Confusing? Just read the last paragraph again and check the respective toString() implementation.
+   * </p>
+   * <p>
+   * This means, in effect, that logging would produce a usable output even if an ordinary System.out.println(o)
+   * would produce a relatively hard-to-debug StackOverflowError.
+   * </p>
+   *
+   * @param o      the Object to convert into a String
+   * @param str    the StringBuilder that o will be appended to
+   * @param dejaVu a list of container identities that were already used.
+   */
+  private static void recursiveDeepToString(@Nullable final Object o, final StringBuilder str, final Set<String> dejaVu) {
+    if (appendSpecialTypes(o, str)) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'appendSpecialTypes' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=165&issues=AWM_D9R25HoPdzsInowa&open=AWM_D9R25HoPdzsInowa)

[ISSUE_KEY]: AWM_D9R25HoPdzsInowa